### ISSUE TITLE
Is there a way to get form input as a dictionary

### ISSUE THREAD

**Issue by InstinctBas** [SergioBenitez/Rocket#951]
I have a form that is constructed based on a user configuring a list of questions to ask. I now want to save the answer the end user fills in on this form on submit into a database.

The problem that I have is that what is being received is fully dependent on the questions configured in the first place so it is not a static structure. Is there a way for form data given to a route to be in the form of a dictionary or an array of name/value pairs?

**Comment by jebrosen** [comment]
It sounds like you want to parse a form whose possible fields are determined at runtime, or to have a freeform type like a `HashMap<String, String>`. The first is relatively difficult, and Rocket doesn't provide the second itself. However, you can create your own type and implement [`FromForm`](https://api.rocket.rs/v0.4/rocket/request/trait.FromForm.html) manually -- with full access to the incoming form fields.

**Comment by InstinctBas** [comment]
Thanks Jeb,

Thats is exactly so, I had no problems creating the form, I'm rebuilding existing logic that is currently embedded in a fat client online. So I simply select the list of questions from the database and provide that to my tera template and it renders it out perfectly.

It is the parsing of what comes back that I'm trying to figure out.
I'm going to look into writing my own FromForm, sounds like that will allow me to do what I need to do.
Interestingly the FormItems that is given to from_form is pretty much what I'm after as an input.

I'll be sure to post what I come up with here :)

**Comment by InstinctBas** [comment]
Hey Jeb,

I came up with this:
```
use rocket::request::{FromForm, FormItems};
use std::collections::HashMap;

pub struct FormDictionary {
	pub fields: HashMap<String, String>
}

impl<'f> FromForm<'f> for FormDictionary {
	type Error = ();

	fn from_form(items: &mut FormItems<'f>, strict: bool) -> Result<FormDictionary, ()> {
		let mut fields = HashMap::new();

		for item in items {
			let decoded_value = item.value.url_decode().unwrap();
			fields.insert(item.key.to_string(), decoded_value);
		}

		Ok(FormDictionary { fields: fields })
	}
}
```

Seems to work :)

**Comment by jebrosen** [comment]
That's about what I imagined but didn't have time to type out and test earlier.

This can be improved by a) handling `_method`, b) don't use `unwrap` and return some kind of `Error` instead, and c) if at all possible, enforce some kind of structure to the form, for example separating fields of the form "answer1", "answer2", etc. from general fields such as "date" or "username". I have no idea what your specific requirements are, but I wanted to mention it for anyone else who comes across this as well -- a form that only exists as a key-value mapping is not very pleasant to work with.

---

Closing because it looks like this is pretty much solved.
